parser/parser.coffee

```diff
@@ -0,0 +1,118 @@
+scanner = require '../scanner/scanner'
+CustomError = require '../error/custom_error'
+
+Program = require '../entities/01_program'
+Block = require '../entities/02_block'
+ForStatement = require '../entities/03_for_statement'
+WhileStatement = require '../entities/03_while_statement'
+
+StartTokens = require './start_tokens'
+
+tokens = []
+
+module.exports = (scannerOutput) ->
+  tokens = scannerOutput
+  program = parseProgram()
+  match 'EOF'
+
+parseProgram = ->
+  new Program parseBlock
+
+parseBlock = ->
+  statements = []
+  loop
+    statements.push parseStatement()
+    match 'newline'
+    break unless at StartTokens.statement
+  new Block statements
+
+parseStatement = ->
+  if at 'for'
+    parseForLoop()
+  else if at 'while'
+    parseWhileLoop()
+  else
+    parseExpression()
+
+parseForLoop = ->
+  match 'for'
+  id = parseId()
```

Comment:
The intent was that I got into a grove and wasn't thinking clearly. I knew that I wouldn't be able to work on this at all this week, so I figured I'd push what code I had (that I knew was broken), so that way if someone else wanted to continue the work, they could build on it.